pysipp/launch.py

```diff
@@ -185,3 +180,14 @@ async def finalize(runner, agents, timeout):
         raise SIPpFailure(msg)
 
     return cmds2procs
+
+
+def finalizer(finalize_coro, runner, agents):
```

Comment:
So if we go with this I guess it's a breaking change right?
We might need a major version release in that case, which is cool, and probably not that cray since we already are cutting py2.

Reply:
do you mean bumping to version 2.0 some weeks after version 1.0 was released? 😄

Reply:
Hehe yeah kinda. 

I have a feeling we're going to break the public API a lot with all the python 3 / async stuff.

I think anyone afraid of that should stick to the `0.1.0` (which i'm sure lots of peeps are using in production already) and then we can move forward making the api more functional and async in a `1.0`

What you think?

Also so sorry for 👻 ing; it's been quite the week 😂

Reply:
ah i just realized the release version you did a while back was 0.1.0 not 1.0.0

Reply:
`finalize_coro` here is what?
If it's an async function i don't think we should have `_coro`.
Also in `trio` in general there's a distance kept from referring to *coroutine* objects directly so I'm thinking we can do a better name here?

Reply:
oh yes, i think we don't even need to pass the finalize there, so we can remove the argument altogether.

Reply:
Actually further, unless I'm missing something, I don't think this is going to work.

If you return the `with_timeout` closure then `trio.run()` isn't going to be invoked until the user *calls* `finalizer()` in which case the when `block=False` here:
https://github.com/SIPp/pysipp/pull/65/files#diff-2b80e3f62bb560133d08a5a07a74bc8eac811db3cdf2680136050b78f37d324dR161

there won't be any background code run. 
I can't remember exactly if we're calling whatever `run_all_agents()` returns but if it does, we should change the name I think?

Reply:
I've removed the argument.
I'm not sure what you mean, `Scenario` runs `TrioRunner`, through `run_all_agents`, which uses `open_subprocess`

Reply:
I'm confused too 😂 

I don't get how this `finalizer()` can work.
Why does it need an inner closure like this?

Isn't `trio.run()` already running at this point, and if so why are we calling it again?

Maybe i'm forgetting how things work.

Reply:
we `trio.run` once to run the subprocesses then once again to wait them to end or until timeout.